src/main.vala

```diff
@@ -16,171 +16,230 @@
-
-                if (variable.auto) {
-                    user_input = "";
-                } else if ((user_input = stdin.read_line ()) == null) {
-                    throw new Valdo.TemplateApplicationError.USER_QUIT ("User has quit");
-                }
+    /**
+     * Application entry point
+     *
+     * @param args command-line arguments
+     *
+     * @return result code
+     */
+    int main (string[] args) {
+        APP_NAME = args[0];
+
+        /* Parse command-line options */
+        var ctx = new OptionContext ("- create a Vala project from a template");
+
+        ctx.set_summary (@"Run $(APP_NAME) without any args to list all available templates");
+        ctx.set_description ("Report bugs to https://github.com/Prince781/valdo/issues");
+        ctx.add_main_entries (ENTRIES, null);
+
+        try {
+            ctx.parse (ref args);
+        } catch (Error e) {
+            stderr.printf ("%s\n", e.message);
+            stderr.printf ("Try '%s --help' for more information\n", APP_NAME);
+            return 1;
+        }
 
-                if (user_input == "" && default_value == null) {
-                    stderr.printf ("Error: %s was not specified\n", variable.summary);
-                } else if (user_input != null) {
-                    if (user_input == "")
-                        user_input = default_value;
-
-                    // verify input
-                    if (variable.pattern != null) {
-                        if (!new Regex (/* FIXME: non-null */(!)variable.pattern).match (/* FIXME: non-null */(!)user_input)) {
-                            stderr.printf ("Error: your entry must match the pattern: %s\n", /* FIXME: non-null */(!)variable.pattern);
-                        } else {
-                            input_verified = true;
-                        }
-                    } else {
-                        input_verified = true;
-                    }
-                }
-            } while (!input_verified);
+        /* --version/-v */
+        if (option_version) {
+            stdout.printf ("%s %s\n", APP_NAME, Config.VERSION);
+            return 0;
+        }
 
-            substitutions[variable.name] = /* FIXME: non-null */ (!)user_input;
+        /* -l/--list-templates */
+        if (option_list_templates) {
+            list_templates ();
+            return 0;
         }
 
-        // now apply the template to the new directory
-        string project_name = substitutions["PROJECT_DIR"];
-        Valdo.TemplateEngine.apply_template (
-            template,
-            File.new_for_path (Environment.get_current_dir ()),
-            project_name,
-            substitutions
-        );
-    } catch (Error e) {
-        if (!(e is Valdo.TemplateApplicationError.USER_QUIT)) {
-            stderr.printf ("Applying template failed\n");
-            stderr.printf ("%s\n", e.message);
+        /* Quit if not one template specified */
+        if (non_option_arguments.length != 1) {
```

Comment:
You should print out all available templates if no arguments are provided.

Reply:
Fixed